File: 345. Reverse Vowels of a String/main.py
class Solution:
    def reverseVowels(self, s: str) -> str:
        def isEmpty(c: chr) -> bool:
            return c in "aeiouAEIOU"

        t = list(s)
        size = len(t)
        left = 0
        right = size - 1

        while left < right:
            while left < size and not isEmpty(t[left]):
                left += 1

            while right >= 0 and not isEmpty(t[right]):
                right -= 1

            if left < right:
                t[left], t[right] = t[right], t[left]
                left += 1
                right -= 1

        return "".join(t)


class Solution:
    def reverseVowels(self, s: str) -> str:
        def isEmpty(c: chr) -> bool:
            return c in "aeiouAEIOU"

        t = list(s)
        size = len(t)
        left = 0
        right = size - 1

        while left < right:
            if not isEmpty(t[left]):
                left += 1

            if not isEmpty(t[right]):
                right -= 1

            if isEmpty(t[left]) and isEmpty(t[right]):
                t[left], t[right] = t[right], t[left]
                left += 1
                right -= 1

        return "".join(t)


class Solution:
    def reverseVowels(self, s: str) -> str:
        vowels = "aeiouAEIOU"
        t = list(s)
        size = len(t)
        left = 0
        right = size - 1

        while left < right:
            if t[left] not in vowels:
                left += 1

            if t[right] not in vowels:
                right -= 1

            if t[left] in vowels and t[right] in vowels:
                t[left], t[right] = t[right], t[left]
                left += 1
                right -= 1

        return "".join(t)

File: 345. Reverse Vowels of a String/main.rs
struct Solution;

impl Solution {
    pub fn reverse_vowels(s: String) -> String {
        let mut t = s.chars().collect::<Vec<_>>();
        let vowels = "aeiouAEIOU";
        let size = t.len();
        let mut left = 0;
        let mut right = size as i32 - 1;

        while left < right {
            while left < size as i32 && !vowels.contains(t[left as usize]) {
                left += 1;
            }

            while right >= 0 && !vowels.contains(t[right as usize]) {
                right -= 1;
            }

            if left < right {
                t.swap(left as usize, right as usize);
                left += 1;
                right -= 1;
            }
        }

        t.iter().collect::<String>()
    }
}

impl Solution {
    pub fn reverse_vowels(s: String) -> String {
        let mut t = s.chars().collect::<Vec<_>>();
        let vowels = "aeiouAEIOU";
        let size = t.len();
        let mut left = 0;
        let mut right = size as i32 - 1;

        while left < right {
            if !vowels.contains(t[left as usize]) {
                left += 1;
            }

            if !vowels.contains(t[right as usize]) {
                right -= 1;
            }

            if vowels.contains(t[left as usize]) && vowels.contains(t[right as usize]) {
                t.swap(left as usize, right as usize);
                left += 1;
                right -= 1;
            }
        }

        t.iter().collect::<String>()
    }
}

impl Solution {
    pub fn reverse_vowels(s: String) -> String {
        let vowels = "aeiouAEIOU";
        let mut r = s.chars().rev().filter(|&c| vowels.contains(c));

        s.chars()
            .map(|c| {
                if vowels.contains(c) {
                    r.next().unwrap()
                } else {
                    c
                }
            })
            .collect()
    }
}

impl Solution {
    pub fn reverse_vowels(s: String) -> String {
        let is_vowels = |c| matches!(c, 'a' | 'e' | 'i' | 'o' | 'u' | 'A' | 'E' | 'I' | 'O' | 'U');
        let mut r = s.chars().rev().filter(|&c| is_vowels(c));

        s.chars()
            .map(|c| if is_vowels(c) { r.next().unwrap() } else { c })
            .collect()
    }
}
